### DISCUSSION TITLE
WSL2 Ubuntu pyenv location문제

### DISCUSSION THREAD

**Discussion by TeusEE** [lablup/contribution-academy-2021#23] [Q&A] [answered]
[backend.ai.txt](https://github.com/lablup/contribution-academy-2021/files/6957942/backend.ai.txt)

Describe the problem
정상적으로 install-dev.sh완료 후

verify installation을 진행하기 위해서 

[NOTE] How to run Backend.AI manager:
> cd /home/xodn/backend.ai-dev/manager
> python -m ai.backend.manager.server --debug

를 진행할 경우 
/usr/bin/python: Error while finding module specification for 'ai.backend.manager.server' (ModuleNotFoundError: No module named 'ai')
에러 발생

추정으로는 pyenv가 기본 directory인 $HOME/.pyenv/bin/pyenv가 아니라
/root/.pyenv/bin/pyenv에 설치되어 자동으로 가상화 환경이 적용이 안되는것으로 보입니다.

![image](https://user-images.githubusercontent.com/79349894/128787568-ae4b20b6-41a4-4296-85f2-60997522b986.png)

Expected behavior
/root/.pyenv 를 기준으로 pyenv가 적용되서 정상적인 verify installation 가능

To Reproduce(WSL 기준)

install python3.9.5 using tar file
install pyenv
install scripts/install-dev.sh
installation complete
when verify installation, pyenv not working(local python activate first)

Screenshots
![image](https://user-images.githubusercontent.com/79349894/128788067-eb47a2f5-c037-4d37-87a0-451051c287ab.png)

OS:  windows10(21H1, 19043.1110) WSL2.0
Current Shell : bash

**Comment by agatha197** [comment]
`echo $0` 했을 때 결과가 `-zsh` 라면 ~/.zshrc 에 다음을 추가하신 후, `source ~/.zshrc` 로 활성화 해주세요. 
```
export PYENV_ROOT="$HOME/.pyenv"
export PATH="$PYENV_ROOT/bin:$PATH"
eval "$(pyenv init --path)"
eval "$(pyenv virtualenv-init -)"
```

**Reply by TeusEE** [reply]
말씀해주신걸로 확인했을때, shell은 bash쉘로 나오고
정상적으로 root가 설정된 것으로 확인됩니다
========================shell scripts=========================
[bashscripts.txt](https://github.com/lablup/contribution-academy-2021/files/6967631/bashscripts.txt)
========================shell scripts=========================
 source ~/.bashrc
이후에 pyenv명령어를 통해서 정보를 찾아보면 아래와같이 나옵니다
![image](https://user-images.githubusercontent.com/79349894/129013228-1b1d0c8a-82df-4c11-a016-7840a98bd9c4.png)


이슈에 명시하긴 했는데, 보시면 pyenv가 $HOME/.pyenv(backend.ai 설치 이전에 python3.9.5와 같이 설치)와 /root/.pyenv(install-dev.sh진행하면서 설치)두개가 있습니다. 해당 원인으로 문제가 발생하는거 같은데, 혹이 어떻게 해야할까요?


두가지로 설치된 pyenv를 확있을을때, 각 directory의 상황은 아래와같습니다
![image](https://user-images.githubusercontent.com/79349894/129012128-6a612643-bc19-46bf-bd6f-5deb1e0d3053.png)

**Reply by agatha197** [reply]
네 확인해보니 언급해주신 원인이 맞는 것 같습니다.
https://github.com/lablup/backend.ai/blob/main/scripts/install-dev.sh#L467 를 보시면 알 수 있듯이, `PYENV_ROOT`의 경로가 `$HOME/.pyenv` 로 되어 있기 때문에 `/root/.pyenv` 을 제거하신 후 재진행해주시기 바랍니다.

**Reply by TeusEE** [reply]
WSL2 Ubuntu20.04 환경에서 install-dev.sh를 진행할 경우,
script와 다르게 /root/.pyenv를 설치위치로 가져가는것 같습니다.
```
[INFO] Installing pyenv...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   270  100   270    0     0    169      0  0:00:01  0:00:01 --:--:--   169
Cloning into '/root/.pyenv'...
remote: Enumerating objects: 773, done.
remote: Counting objects: 100% (773/773), done.
remote: Compressing objects: 100% (403/403), done.
remote: Total 773 (delta 424), reused 482 (delta 275), pack-reused 0
Receiving objects: 100% (773/773), 412.78 KiB | 2.01 MiB/s, done.
Resolving deltas: 100% (424/424), done.
Cloning into '/root/.pyenv/plugins/pyenv-doctor'...
remote: Enumerating objects: 11, done.
remote: Counting objects: 100% (11/11), done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 11 (delta 1), reused 5 (delta 0), pack-reused 0
Unpacking objects: 100% (11/11), 38.62 KiB | 205.00 KiB/s, done.
Cloning into '/root/.pyenv/plugins/pyenv-installer'...
remote: Enumerating objects: 16, done.
remote: Counting objects: 100% (16/16), done.
remote: Compressing objects: 100% (13/13), done.
remote: Total 16 (delta 1), reused 9 (delta 0), pack-reused 0
Unpacking objects: 100% (16/16), 5.81 KiB | 991.00 KiB/s, done.
Cloning into '/root/.pyenv/plugins/pyenv-update'...
remote: Enumerating objects: 10, done.
remote: Counting objects: 100% (10/10), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 10 (delta 1), reused 5 (delta 0), pack-reused 0
Unpacking objects: 100% (10/10), 2.83 KiB | 965.00 KiB/s, done.
Cloning into '/root/.pyenv/plugins/pyenv-virtualenv'...
remote: Enumerating objects: 57, done.
remote: Counting objects: 100% (57/57), done.
remote: Compressing objects: 100% (51/51), done.
remote: Total 57 (delta 11), reused 22 (delta 0), pack-reused 0
Unpacking objects: 100% (57/57), 35.37 KiB | 1.41 MiB/s, done.
Cloning into '/root/.pyenv/plugins/pyenv-which-ext'...
remote: Enumerating objects: 10, done.
remote: Counting objects: 100% (10/10), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 10 (delta 1), reused 6 (delta 0), pack-reused 0
Unpacking objects: 100% (10/10), 2.92 KiB | 995.00 KiB/s, done.

WARNING: seems you still have not added 'pyenv' to the load path.
```
재현성 
Ubuntu, Docker초기화 -> Python3.9 설치 -> install-dev.sh 순으로 진행 하였습니다.

install 완료 메세지를 확인 하였으나, 아래 스크린샷처럼 /root/.pyenv만 존재하고, $HOME/.pyenv를 생성되지가 않습니다
![image](https://user-images.githubusercontent.com/79349894/129438562-2c7470ae-b3c3-425d-8ff0-a4d94fa8d189.png)

**Reply by agatha197** [reply]
Ubuntu 에서 bash를 사용하실 경우 `~/.bashrc` 에 다음과 같이 설정하시기 바랍니다.
```
export PATH="/home/<username>/.pyenv/bin:$PATH"
export PATH="$PYENV_ROOT/bin:$PATH"
eval "$(pyenv init -)"
eval "$(pyenv virtualenv-init -)"
```
[관련 링크](https://github.com/pyenv/pyenv-installer/issues/103)

---

WSL2.0 Ubuntu 18.04 LTS 로 설치 성공하신 분이 있으니 Ubuntu 18.04 LTS에서 시도해보셔도 좋을 것 같습니다.

**Comment by VSFe** [comment]
이전에 올린 글을 보니 pyenv 검증 과정을 회피해서 (SSL 검증. pyenv가 비정상적으로 동작하는 상황에서 python 실행 시 python2가 실행되어 문제 발생) installation을 진행한 것 같습니다.

```
echo 'eval "$(pyenv init -)"' >> ~/.bashrc
source ~/.bashrc
```

를 입력하고 다시 설치 진행 해보시겠어요?

**Reply by VSFe** [reply]
기존 스크립트엔 `eval "$(pyenv init -)"` 구문이 포함되어 있었는데, 최근에 수정을 진행하면서 빠진 것 같습니다.

저 또한 WSL2 환경을 사용해서 설치했었는데, 저 구문을 포함해서 설치를 진행하니 성공적으로 구동되었습니다.

**Comment by TeusEE** [comment, marked as answer]
문제 해결했습니다.
확인해보니깐

install-dev.sh shell scripts를 실행할 때 실행 중에는
1. WSL2 Ubuntu 18.04 => sudo backend.ai/scripts/install-dev.sh 실행 시 $HOME = /home/{user_nam}
2. WSL2 Ubuntu 20.04 => sudo backend.ai/scripts/install-dev.sh 실행 시 $HOME = /root
로 고정되어서 문제가 발생했습니다.

sudo 없이 일반유저권한으로 실행하여 문제 해결되었습니다.
